<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Eff3ctidor</title>

    <link rel="shortcut icon" type="image/x-icon" href="/pics/logo.gif" />

    <meta charset="utf-8">
    <!--<meta name="viewport" content="width=device-width, initial-scale=1">-->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
    <link rel="stylesheet" href="https://gitcdn.github.io/bootstrap-toggle/2.2.2/css/bootstrap-toggle.min.css">
    
    <!--files for the select tag-->
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/css/bootstrap-select.min.css">
    <script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/js/bootstrap-select.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-select/1.12.4/js/i18n/defaults-*.min.js"></script>

    <!--files for the checkbox tag (toggle)-->
    <script src="https://gitcdn.github.io/bootstrap-toggle/2.2.2/js/bootstrap-toggle.min.js"></script>

    <link rel="stylesheet" href="css/general.css">
    <script src="handle_FAQ.js"></script>
</head>
<body>
    <nav role="navigation" class="navbar navbar-fixed-top">
        <div class="jumbotron" id="jumbo">
            <div class="container">            
                <div class="row" id="title-row" align="center">
                    <div class="col-md-1">
                    </div>
                    <div class="col-md-10" align="center">
                        <!--<span id="server-title">Effectidor</span><br>-->
                        <img src="pics/logo.gif" id="nav_bar_image" style="height: 120px;"><br>
                        <span id="sub-title">A machine learning based type 3 effectors predictor</span>
                        <br><br>
                    </div>
                </div>
            </div>       
        </div>
        <div class="jumbotron" id="nav_container">
                <div class="row">
                    <div style="display:table; margin:0 auto;">
                        <ul class="nav navbar-nav" id="menu-nav">
                            <li style="display:table-cell; padding: 0px 10px;">
                                <a href="index.html">Home</a>
                            </li>
                            <li class="active" style="display:table-cell; padding: 0px 10px;">
                                <a href="FAQ.html">FAQ</a>
                            </li>
                            <li style="display:table-cell; padding: 0px 10px;">
                                <a href="data.html">Data</a>
                            </li>
                            <li style="float: right">
                                <a href="https://english.tau.ac.il/" id="tau-link" target="_blank">
                                    <img src=pics/TAU_LOGO.png id="tau-logo">
                                </a>
                            </li>
                        </ul>
                    </div>
                </div>
        </div>
    </nav>
    <div id="behind-nav-bar-results"></div>
    <div class="container">
    <h3 onclick="Open_output()" style="cursor:pointer;">What is Effectidor's output?</h3><br>
    <div id="outputQ" style="display:none;">
    <p>Effectidor produces both tabular and graphical outputs.
    In the tabular output there are:<br></p>
    <ol>
        <li>The T3Es' predictions, i.e., each ORF gets a score that reflects its likelihood to encode a type 3 effector, as produced by the machine-learning.
    In this file you will have the locus_tags sorted by the ORFs' likelihood to encode effectors. The protein sequence will appear next to it, in addition to the protein annotation, if was available in the input data.
    <br>The 10 best scoring, unlabeled predictions, as well as the positive set that was used to train the classifier, will be displayed on the screen, while the predictions for all the ORFs in the genome (labeled data as well as unlabeled data) will be available in a downloadable Excel file.</li>
        <li>Pseudogenes file - an Excel file with all the predictions of ORFs that are annotated as pseudogenes.
        <br>Highly scored samples available in this file are potentially with effector origin that was lost in evolution.
        <br>This file will be empty if the input ORFs file did not contain annotations.</li>
        <li>Feature importance table. This table shows the features by their contribution to the learning, and their relative importnce. It is available as a downloadable csv file.</li>
        <li>Raw features file. The features file, containing all the features that were extracted from the input data, that was used to train the machine learning algorithm, is also available as a downloadable csv file.</li>
    </ol>
    <p>
    The graphical output shows analysis of the features. The 10 most contributing features will be compared between the two classes - effectors and non-effectors, in violin plots.
    <br><br>You can see an example for the output in the running example <a href="/results/example/output.html" target="_blank">here</a>.
    </p><br>
    </div>
    <h3 onclick="Open_interpretation()" style="cursor:pointer;">How do I interpret the results?</h3><br>
    <div id="interpretationQ" style="display:none;">
    <p>The main results, the predictions file, contains the ORFs sorted by their likelihood to encode T3Es.
    <br>To identify candidates to be novel T3Es, look at the top rows of the table, and search ORFs with a high score.
    <br>If you used an input Effectors file, look for samples that are not already known to be effectors. These samples will have a question mark (?) in the column "is_effector".
    <br>If you have not used an input Effectors file, The samples marked with "yes" in the "is_effector" column are homologous to known T3Es, and the highly scored samples marked with a question mark (?) are potentially novel T3Es.
    <br>For your convenience, the highly scored samples are highlighted with dark green that becomes lighter as the score decreases.
    <br>If no candidate is found at the top of the file, you can carry on and check lower rows in this file.
    <br>For your convenience, the amino acid sequence will be available to each ORF, as well as the annotation (if existed in the input file).
    <br><br>In general, an ORF with a high score and an ambiguous annotation is a good candidate to be a T3E.</p>
    <br>
    </div>
    <h3 onclick="Open_evaluation()" style="cursor:pointer;">How is Effectidor evaluated and where can I see its predicting accuracy?</h3><br>
    <div id="evaluationQ" style="display:none;">
    <p>In the learning process Effectidor randomly leaves 20% of the labeled data (effectors and non-effectors) aside, as a test set. Each learning algorithm is trained on the remaining 80% training data, using cross validation, and in the end it is evaluated on the untouched 20% test set.
    <br>The measurement used to evaluate Effectidor's accuracy is the Area Under the Precision Recall Curve (AUPRC). The closer it is to 1, the more accurate the classifier is. You can see the resulting AUPRC achieved on the test set, in the downloadable predictions file.
    </p><br>
    </div>
    <h3 onclick="Open_input()" style="cursor:pointer;">What input does Effectidor require and what do they use for?</h3><br>
    <div id="inputQ" style="display:none;">
    <p>
    Effectidor has one obligatory input - <b>ORFs file</b>.<br> This file is a FASTA file with all the ORFs in the genome. See instructions on how to download this file <a href="instructions.html" target="_blank">here</a>.
    <br>Some of the ORFs in this file (effectors and non-effectors) will be used to train the machine-learning algorithms, and based on the trained classifier, the main output - prediction for each ORF - will be performed. 
    <br>In addition, it is recommended to supply an <b>Effectors file</b>. This file should contain all the known effectors' rocords, as appear in the ORFs file. If you do not supply such a file, a homology search against an internal effectors dataset will be performed to constitute the "known effectors" ORFs for the learning process. 
    <br><br> In the advanced options you can supply additional data that can improve the analysis. Each of these additional data will result in additional features that will feed the machine-learning.<br>These data include:</p>
    <ol>
        <li><b>Host proteome</b> archive. Protein <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA</a> files with the proteome of a known host of the studied bacterium. Multiple files can be included. All these files should be compressed together in a single <a href="https://support.microsoft.com/en-us/windows/zip-and-unzip-files-8d28fa72-f2f9-712f-67df-f80cf89fd4e5" target="_blank">zip</a> archive.
        <br>This input will be used for homology searches. As effectors interact with host proteins for their function, we expect them to show homology to host proteins.
        </li>
        <li>Archive of <b>proteomes of closely related bacteria without T3SSs</b>. This archive may contain several proteomes records, each in a different <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA</a> file. These FASTA files should be compressed together in a single <a href="https://support.microsoft.com/en-us/windows/zip-and-unzip-files-8d28fa72-f2f9-712f-67df-f80cf89fd4e5" target="_blank">zip</a> archive.
        <br>Against each of these proteomes a homology search will be performed. As these bacteria are closely related to the studied bacterium, the vast majority of the proteins in the studied bacterium are expected to have an ortholog in these proteomes. Nevertheless, since they do not encode a T3SS, effectors are not expected to have orthologs in these proteomes. Thus, these features are usually very informative for the machine-learning.
        </li>
        <li>
        <b>Fully assembled genome</b> question box. By default this box is marked with "no". If the genome you are analyzing is composed of one circular contig, change it to "yes". It will allow calculating features of genome organization.
        </li>
        <li>
        <b><a href="http://gmod.org/wiki/GFF3" target="_blank">GFF3</a> and full genome <a href="https://www.ncbi.nlm.nih.gov/blast/fasta.shtml" target="_blank">FASTA</a> files</b>. These input files should be supplied together, and only if the genome is fully assembled (see previous input) and if the bacterium you study is a phytopathogenic bacteria. If this is the case and you supply these data, several known motifs that known TFs of T3Es bind to will be searched in the promoters of the genes.
        </li>
    </ol>
    </p>
    </div>
    <h3 onclick="Open_running_time()" style="cursor:pointer;">What is the expected running time of Effectidor?</h3><br>
    <div id="running_timeQ" style="display:none;">
    <p>The running time depends on many factors, among them the input you supply and the load in our servers.
    <br>It can take a few minutes and up to several hours. We will send you an email with a link to the results upon submission, if you choose to supply your email address.
    </p><br>
    </div>
    <h3 onclick="Open_saving_time()" style="cursor:pointer;">For how long will my results be saved in the servers?</h3><br>
    <div id="saving_timeQ" style="display:none;">
    <p>The results will be saved in our servers for 3 months.
    <br>After 3 months they will be permanently deleted from our servers.</p><br>
    </div>
    <h3 onclick="Open_email()" style="cursor:pointer;">Why do you need my email?</h3><br>
    <div id="emailQ" style="display:none;">
    <p>If you give us your email upon submission we will use it to send you a link to your results, and nothing more.<br>
    It is an optional input and you do not have to give us your email in order to use Effectidor.</p><br>
    </div>
    <h3 onclick="Open_genomes()" style="cursor:pointer;">Can I use Effectidor to run the analysis on several genomes simultaneously?</h3><br>
    <div id="genomesQ" style="display:none;">
    <p>As of today, Effectidor only works for one genome at a time.
    <br>You can submit a different job for each genome, but you cannot run a common analysis for more than one genome.
    <br>This may change in the future, stay tuned.</p>
    </div>
    <div id="behind-footer"></div>
    </container>
    <nav role="navigation" class="navbar-inverse navbar-fixed-bottom ng-binding" style="color: #cccccc; text-align: center; font-family: Calibri;">
        <h4>
                Research Site | <a href="https://www.tau.ac.il/~talp/" target="_blank">Pupko Group</a>
        </h4>
    </nav>
</doby>
</html>